mongo/src/test/scala/io/sphere/mongo/generic/MongoEmbeddedSpec.scala

```diff
@@ -0,0 +1,150 @@
+     value1: String,
+     @MongoKey("_value2") value2: Int)
+
+  object Embedded {
+    implicit val mongo: MongoFormat[Embedded] = mongoProduct(apply _)
+  }
+
+  case class Test1(
+    name: String,
+    @MongoEmbedded embedded: Embedded)
+
+  object Test1 {
+    implicit val mongo: MongoFormat[Test1] = mongoProduct(apply _)
+  }
+
+  case class Test2(
+    name: String,
+    @MongoEmbedded embedded: Option[Embedded] = None)
+
+  object Test2 {
+    implicit val mongo: MongoFormat[Test2] = mongoProduct(apply _)
+  }
+
+  case class Test3(
+    @MongoIgnore name: String = "default",
+    @MongoEmbedded embedded: Option[Embedded] = None)
+
+  object Test3 {
+    implicit val mongo: MongoFormat[Test3] = mongoProduct(apply _)
+  }
+
+  case class SubTest4(@MongoEmbedded embedded: Embedded)
+  object SubTest4 {
+    implicit val mongo: MongoFormat[SubTest4] = mongoProduct(apply _)
+  }
+
+  case class Test4(subField: Option[SubTest4] = None)
+  object Test4 {
+    implicit val mongo: MongoFormat[Test4] = mongoProduct(apply _)
+  }
+}
+
+class MongoEmbeddedSpec extends WordSpec with MustMatchers with OptionValues {
+  import MongoEmbeddedSpec._
+
+  "MongoEmbedded" should {
+    "flatten the db object in one object" in {
+      val dbo = dbObj(
+        "name" -> "ze name",
+        "value1" -> "ze value1",
+        "_value2" -> 45
+      )
+      val test1 = MongoFormat[Test1].fromMongoValue(dbo)
+      test1.name mustEqual "ze name"
+      test1.embedded.value1 mustEqual "ze value1"
+      test1.embedded.value2 mustEqual 45
+
+      val result = MongoFormat[Test1].toMongoValue(test1)
+      result mustEqual dbo
+    }
+
+    "validate that the db object contains all needed fields" in {
+      val dbo = dbObj(
+        "name" -> "ze name",
+        "value1" -> "ze value1"
+      )
+      Try(MongoFormat[Test1].fromMongoValue(dbo)).isFailure must be (true)
+    }
+
+    "support optional embedded attribute" in {
+      val dbo = dbObj(
+        "name" -> "ze name",
+        "value1" -> "ze value1",
+        "_value2" -> 45
+      )
+      val test2 = MongoFormat[Test2].fromMongoValue(dbo)
+      test2.name mustEqual "ze name"
+      test2.embedded.value.value1 mustEqual "ze value1"
+      test2.embedded.value.value2 mustEqual 45
+
+      val result = MongoFormat[Test2].toMongoValue(test2)
+      result mustEqual dbo
+    }
+
+    "ignore unknown fields" in {
+      val dbo = dbObj(
+        "name" -> "ze name",
+        "value1" -> "ze value1",
+        "_value2" -> 45,
+        "value4" -> true
+      )
+      val test2 = MongoFormat[Test2].fromMongoValue(dbo)
+      test2.name mustEqual "ze name"
+      test2.embedded.value.value1 mustEqual "ze value1"
+      test2.embedded.value.value2 mustEqual 45
+    }
+
+    "ignore ignored fields" in {
+      val dbo = dbObj(
+        "value1" -> "ze value1",
+        "_value2" -> 45
+      )
+      val test3 = MongoFormat[Test3].fromMongoValue(dbo)
+      test3.name mustEqual "default"
```

Comment:
Im slightly confused here. I would expect the scala representation to have the field `name` and that it is dropped in `MongoFormat[Test3].toMongoValue(test3)` but how come the `name` field is added in `test3` when it isn't in the mongo obj?

Reply:
Because `Test3` defines a default value for `name`:
```
  case class Test3(
    @MongoIgnore name: String = "default",
```

There was no test for that until now 😭